src/Core/Context/Data.idr

```diff
@@ -92,20 +92,21 @@ paramPos tyn dcons = do
 export
 addData : {auto c : Ref Ctxt Defs} ->
           Scope -> Visibility -> Int -> DataDef -> Core Int
-addData vars vis tidx (MkData (MkCon dfc tyn arity tycon) datacons)
+addData vars vis tidx (MkData con datacons)
     = do defs <- get Ctxt
-         let allPos = NatSet.allLessThan arity
+         let tyName = con.name.val
+         let allPos = NatSet.allLessThan con.arity
          -- In case there are no constructors, all the positions are parameter positions!
-         let paramPositions = fromMaybe allPos !(paramPos (Resolved tidx) (map type datacons))
+         let paramPositions = fromMaybe allPos !(paramPos (Resolved tidx) (map val datacons))
          log "declare.data.parameters" 20 $
-            "Positions of parameters for datatype" ++ show tyn ++
+            "Positions of parameters for datatype" ++ show tyName ++
             ": " ++ show paramPositions
-         let tydef = newDef dfc tyn top vars tycon (specified vis)
-                            (TCon arity
+         let tydef = newDef con.fc tyName top vars con.val (specified vis)
+                            (TCon con.arity
                                   paramPositions
                                   allPos
-                                  defaultFlags [] (Just $ map name datacons) Nothing)
-         (idx, gam') <- addCtxt tyn tydef (gamma defs)
+                                  defaultFlags [] (Just $ map (\x => x.name.val) datacons) Nothing)
```

Comment:
Maybe `map (.name.val) datacons` would be better?

Reply:
Good point! I think I was hitting some ambiguity problems at some point but it seems resolved now. I've also updated other parts of the PR with the same pattern